keps/sig-scheduling/3521-pod-schedule-readiness/README.md

```diff
@@ -0,0 +1,987 @@
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+Pods are naturally perceived ready for scheduling once created. Kubernetes scheduler does
+its due diligence to find nodes to place all pending Pods. However, in a real-world case,
+some Pods may stay in a "miss-essential-resources" state for a long period. These pods,
+actually churn the scheduler (and downstream integrators like Cluster AutoScaler) in an
+unnecessary manner.
+
+Lacking the knob to flag Pods as schedule-ready/unready wastes scheduling cycles on retrying
+Pods that are determined to be unschedulable. As a result, it delays the scheduling of other Pods,
+and also lowers the overall scheduling throughput. Moreover, it imposes restrictions to vendors to
+develop some in-house features (such as hierarchical quota) natively.
+
+This proposal comes up with an extensible mechanics to ensure Pods are scheduled only when they are
+semantically ready.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+- Define API to mark Pod as schedule-paused/unpaused.
+- Design a new Enqueue extension point to customize Pod's queueing behavior.
+- A default enqueue plugin to honor the new API semantics.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+- Update the Pod's condition if it carries `.status.schedulePaused=true`.
+- Focus on in-house use-cases of the Enqueue extension point.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+We propose a new field `.status.schedulePaused` to the Pod API. The field is defaulted to `false`.
+
+For Pods carrying `.status.schedulePause=true`, they will be "parked" in scheduler's internal
+unschedulablePods pool, and only gets retried when the field is set to `false`.
+
+Practically, this field can be initialized to `true` by MutatingWebhook, and be flipped to `false`
+by external integrators when certain criteria is met.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+As an application developer, I'm knowledgeable to know when my Pods are schedule-ready, but I know
+nothing about scheduling internals. It'd be ideal for me to only set/unset the "paused" flag,
```

Comment:
Can I set this value on any pod?  Any workload object with pod templates?  Why or why not?

Reply:
I think in most cases this field will be set to `true` in a pod-level webhook. But I can't think of a reason why we would want to prevent setting it on workload objects.

Reply:
On any pod, yes. PodTemplate, yes (will update the design to add the field in PodSpec so it can be changed in PodTemplate).